### ISSUE TITLE
HLS：同个streamname推流编码x264/h264切换带来的问题

### ISSUE THREAD

**Issue by cnkedao** [ossrs/srs#2206]
**描述(Description)**

> 一个streamname，只要前面使用过x264/h264其中一个编码器，后面换另外一个编码器的时候，HLS只有声音，没有画面(黑屏)

1. SRS版本(Version): `SRS/3.0.141`
1. SRS的日志如下(Log):
```
日志未异常
```
1. SRS的配置如下(Config):
用hls.conf
```
listen              1935;
max_connections     1000;
http_server {
    enabled         on;
    listen          8080;
    dir             ./objs/nginx/html;
}
vhost __defaultVhost__ {
    hls {
        enabled         on;
        hls_path        ./objs/nginx/html;
        hls_fragment    10;
        hls_window      60;
    }
}
```

**重现(Replay)**

> 重现Bug的步骤(How to replay bug?)
假设全程固定使用/live/livestream
1. 先使用OBS，配置H264编码，给/live/livestream进行推流，此时测试HLS正常，然后断开推流
2. 接着使用OBS，配置X264编码，给/live/livestream进行推流，此时测试HLS已经异常，只有声音，没有画面
3. 如果先使用X264推流，然后后面换H264也是一样的问题。整个测试就HLS有问题，RTMP/HTTP-FLV都是正常的
4. 只要是同个streamname，隔了一天再切换编码还是能出现一样的问题

![srs](https://user-images.githubusercontent.com/17036747/108669441-02b4cb80-7518-11eb-9c45-1a42cad0f7c6.jpg)

**期望行为(Expect)**

> 期望SRS能兼容X264/H264切换的情况下，HLS画面都正常。（测试PingOS[ng-rtmp]的HLS是能兼容X264/H264切换的情况也正常）

**Comment by luanche** [comment]
Encountered the same problem, after using x264 encoding, there is only audio but no video when using Nvidia NVENC h.264 (new).

`TRANS_BY_GPT3`

**Comment by luanche** [comment]
> Encountered the same problem, after using x264 encoding, when using Nvidia NVENC h.264 (new), there will still be sound but no video.

> However, some laptops with built-in QuickSync H.264 encoders work fine.

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
👍 I will see if I can reproduce it in my environment.

`TRANS_BY_GPT3`

**Comment by cnkedao** [comment]
> 👍 I will see if I can reproduce it in my environment.

In version srs4.0.71, this issue no longer exists, indicating that there was a fix for this problem in one of the commits in version 4.0. Another issue in version 3.0 is that when recording mp4 files for more than two hours, the mp4 file becomes corrupted (possibly due to an integer overflow) and cannot be played. This issue has also been fixed in version 4.0, as seen in the commit.

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
@cnkedao Thank you for your verification. If version 4.0 has already addressed it, we do not intend to fix it in version 3.0.

`TRANS_BY_GPT3`
